---
title: "Real-World Data Analysis"
author: "Lujun Zhang and Jun Chen"
date: "2024-05-29"
output:
  html_document:
    number_sections: true
---
    
# Introduction and Setups

-  In this document, we will analyze real-world data using DiSC in the dataset:
    +  COVID-PBMC
-  Required packages: 
    +  `doParallel`, `foreach`, `doRNG`, `Matrix`, `matrixStats`, `GUniFrac`, `tidyverse`, `data.table`, `Hmisc`, `kableExtra`, `ggVennDiagram`, `reshape2`, `ggpubr` from CRAN; 
    +  `DESeq2`, `clusterProfiler`, `AnnotationDbi`, `org.Hs.eg.db` from bioconductor; 
    +  `ideas` from [https://github.com/Sun-lab/ideas](https://github.com/Sun-lab/ideas)


```{r setup, include=TRUE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(tidyverse)
library(ggVennDiagram)
library(ggpubr)
library(gridExtra)
library(cowplot)
library(matrixStats)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Hs.eg.db)

competing_methods = c("DiSC", "IDEAS", "DESeq2") # the order in figures. Update this after adding or removing methods
```

# COVID-PBMC

The following code block generates scripts for high-performance computing, which include several R scripts and shell scripts located in the "real_data_anal" folder. These analyses are based on a COVID-PBMC dataset, which was originally obtained by Stephenson et al. (2021) to explore the immune response to COVID-19 using single-cell multi-omics method.

The dataset with raw and normarlized data and metadata can be accessed from https://www.covid19cellatlas.org/, specifically from the link: https://covid19.cog.sanger.ac.uk/submissions/release1/haniffa21.processed.h5ad.

An explanation of the data format has been provided on https://docs.google.com/document/d/1DGlwwft72NkcTdb1NDhb2k-KBljfOnN-CFPkcADy3Uc/edit.

-  Specify the parameters and generate scripts

```{r setting the scripts}
NCORES = 40 %>% as.character()
MEM = 4000 %>% as.character()
WORKDIR = "/home/guanwh/zhan7474/mayo/real_data_anal/"
DATADIR = "/home/guanwh/zhan7474/mayo/real_data_anal/COVID_PBMC"
DATE = "0612" # version control

file.copy(from = "source_code/DiSC.R", to = "real_data_anal/DiSC.R",
          copy.date = TRUE, overwrite  = TRUE)
file.copy(from = "source_code/DiSC_anal.R", 
          to = "real_data_anal/DiSC_anal.R",
          copy.date = TRUE, overwrite  = TRUE)
file.copy(from = "source_code/IDEAS_anal.R", 
          to = "real_data_anal/IDEAS_anal.R",
          copy.date = TRUE, overwrite  = TRUE)
file.copy(from = "source_code/DESeq2_anal.R", 
          to = "real_data_anal/DESeq2_anal.R",
          copy.date = TRUE, overwrite  = TRUE)

file.copy(from = "source_code/COVID_PBMC_subset.sh", 
          to = "real_data_anal/COVID_PBMC_subset.sh",
          copy.date = TRUE, overwrite  = TRUE)
file.copy(from = "source_code/COVID_PBMC_subset.R", 
          to = "real_data_anal/COVID_PBMC_subset.R",
          copy.date = TRUE, overwrite  = TRUE)

read_file("source_code/COVID_PBMC_anal.R") %>%
  str_replace_all(pattern = "DATE", DATE) %>%
  str_replace_all(pattern = "WORKDIR", WORKDIR) %>%
  str_replace_all(pattern = "NCORES", NCORES) %>%
  str_replace_all(pattern = "DATADIR", DATADIR) %>%
  write_file(file = str_glue("real_data_anal/COVID_PBMC_anal_{DATE}.R"), 
             append = FALSE)
read_file("source_code/COVID_PBMC_anal.sh") %>%
  str_replace_all(pattern = "DATE", DATE) %>%
  str_replace_all(pattern = "WORKDIR", WORKDIR) %>%
  str_replace_all(pattern = "NCORES", NCORES) %>%
  str_replace_all(pattern = "MEM", MEM) %>%
  write_file(file = str_glue("real_data_anal/COVID_PBMC_anal_{DATE}.sh"), 
             append = FALSE)
```

-  Some filtering and quality control required for the dataset.

```{bash shell covid pbmc subset, file = "real_data_anal/COVID_PBMC_subset.sh", eval=FALSE}
```

```{r R covid pbmc subset, file = "real_data_anal/COVID_PBMC_subset.R", eval=FALSE}
```

-  Analyses (performed using a high-performance computer)

```{bash shell covid pbmc analysis, file = str_glue("real_data_anal/COVID_PBMC_anal_{DATE}.sh"), eval = FALSE}
```

```{r R covid pbmc analysis, file = str_glue("real_data_anal/COVID_PBMC_anal_{DATE}.R"), eval = FALSE}
```

-  Results

```{r covid-pbmc res, fig.width=8, fig.height=10}
pval_thre = 0.05 # the FDR threshold
#list the results and convert them to a dataframe
res <- list.files("real_data_anal/res/") %>% 
  # should look like "d@COVIDPBMC-m@DESeq2-v@0612-c@B_cell.rds"
  # "varname@value" and splitted by "-" 
  str_subset(pattern = ".rds$") %>%
  str_remove(pattern = ".rds$") %>%
  sapply(X = ., simplify = FALSE, FUN = \(x){
    temp <- str_split(str_split_1(x, "-"), "@", simplify = TRUE)
    return(c(filname = str_glue("{x}.rds"), 
             setNames(object = temp[,2], nm = temp[,1])))
  }) %>% 
  bind_rows()

res <- res %>% 
  filter(d == "COVIDPBMC" & v == DATE) %>%
  dplyr::rename(Method = "m", Cell_Type = "c")
num_sig <- numeric(nrow(res))
name_sig <- character(nrow(res))
for(rr in 1:nrow(res)){
  pval_mat <- readRDS(str_glue("real_data_anal/res/{res$filname[rr]}"))
  sig_genes <- colnames(pval_mat)[pval_mat["adjusted", ] <= pval_thre]
  sig_genes <- sig_genes[!is.na(sig_genes)] # some values can be NAs in DESeq2
  name_sig[rr] <- str_c(sig_genes, collapse = "::")
  num_sig[rr] <- length(sig_genes)
  rm(pval_mat)
}
res <- res %>%
  mutate(num_sig = num_sig, name_sig = name_sig) %>%
  mutate(Method = factor(Method, levels = competing_methods)) %>%
  mutate(Cell_Type = factor(Cell_Type)) %>%
  # order the cell types according to the how many significant genes DiSC found
  mutate(Cell_Type = fct_reorder2(.f = Cell_Type, .x = Method, .y = num_sig, 
                                  .desc = FALSE,
                                  .fun = \(.x, .y) return(.y[.x == "DiSC"])))


# Number of positive findings in the COVID-PBMC dataset
res %>% dplyr::select(Method, Cell_Type, num_sig) %>%
  arrange(desc(Cell_Type), Method) %>% # order a factor
  pivot_wider(names_from = Cell_Type, values_from = num_sig) %>%
  kbl(digits = 4, 
      caption = "Number of Positive Findings in COVID-PBMC dataset(Adjusted for FDR)") %>% 
  kable_classic(full_width = F, html_font = "Cambria")

figure1 <- res %>% 
  ggplot(., aes(x=Cell_Type, color=Method, fill=Method, y=num_sig)) +
  geom_bar(stat = "identity", 
           position = position_dodge2(padding = 0.6, reverse = TRUE), 
           width=0.7) + 
  scale_y_sqrt()+
  xlab("Cell Types") + 
  ylab("Number of DE Genes Detected")+
  theme_bw() + coord_flip() + 
  scale_color_brewer(palette = "Set1") + 
  scale_fill_brewer(palette = "Set1") +
  theme(legend.position='right') 

# Overlaps in positive findings of CD4 and CD8 between methods
Venn_break <- res %>% filter(Cell_Type %in% c("CD4", "CD8"))
Venn_break
Venn_break <- seq(from = 0,
                  to = ceiling(max(Venn_break$num_sig)/100)*100,
                  length.out = 5)
# cell_to_disp <- res %>% 
#   group_by(Cell_Type) %>% 
#   summarise(median_across_method = median(num_sig)) %>%
#   filter(median_across_method == max(median_across_method)) %>% 
#   pull(Cell_Type) %>% as.character()
Venn_CD4 <- res %>% 
  filter(Cell_Type == "CD4") %>%
  mutate(name_sig2 = str_split(name_sig, "::", simplify = FALSE)) %>%
  mutate(name_sig2 = magrittr::set_names(x = name_sig2, value = Method)) %>%
  pull(name_sig2)
Venn_CD4 <- Venn_CD4[base::match(competing_methods, names(Venn_CD4))] #df[match(target, df$name),]
figure2 <- ggVennDiagram(Venn_CD4, label = c("count")) + 
  scale_color_brewer(palette = "Set1") + 
  scale_fill_steps(high = "grey60", low = "grey100", breaks = Venn_break, 
                   limits = c(min(Venn_break), max(Venn_break)),name = "Counts") +
  scale_x_continuous(expand = expansion(mult = .2))+
  theme(legend.position='bottom')
Venn_CD8 <- res %>% 
  filter(Cell_Type == "CD8") %>%
  mutate(name_sig2 = str_split(name_sig, "::", simplify = FALSE)) %>%
  mutate(name_sig2 = magrittr::set_names(x = name_sig2, value = Method)) %>%
  pull(name_sig2)
Venn_CD8 <- Venn_CD8[base::match(competing_methods, names(Venn_CD8))] #df[match(target, df$name),]
figure3 <- ggVennDiagram(Venn_CD8, label = c("count")) + 
  scale_color_brewer(palette = "Set1") + 
  scale_fill_steps(high = "grey60", low = "grey100", breaks = Venn_break, 
                   limits = c(min(Venn_break), max(Venn_break)),name = "Counts") +
  scale_x_continuous(expand = expansion(mult = .2))+
  theme(legend.position='bottom')
#the two figures will have the same legend as specified in 
#breaks = Venn_break and limits = c(min(Venn_break), max(Venn_break))
#but it is still good to check this before align the plots
figure2
figure3
figure2_3 <- lemon::grid_arrange_shared_legend(figure2, figure3, 
                                               nrow = 1, position = "right") %>%
  as_ggplot()


# GO enrichment analysis 
firstup <- function(x) {
  # uppercase the first letter in GO descriptions
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}
deg_CD4 <- res %>% 
  filter(Cell_Type == "CD4", Method == "DiSC") %>%
  pull(name_sig) %>%
  str_split_1(pattern = "::")
deg_CD4_other_methods <- res %>% 
  filter(Cell_Type == "CD4", Method != "DiSC") %>%
  pull(name_sig) %>% 
  str_c(collapse = "::") %>%
  str_split_1(pattern = "::") %>% 
  unique()
go_enrichment_CD4_BP <- enrichGO(gene = deg_CD4, OrgDb = "org.Hs.eg.db", 
                                 keyType = "SYMBOL", ont = "BP") 
go_enrichment_CD4_BP <- go_enrichment_CD4_BP %>% 
  as_tibble() %>% 
  mutate(Count_overlap = sapply(X = geneID, 
                                FUN = \(x, g_vec){
                                  x = str_split_1(x, pattern = "/")
                                  return(sum(x %in% g_vec))}, 
                                simplify = TRUE, USE.NAMES =FALSE,
                                g_vec = deg_CD4_other_methods),
         Count_DiSC = Count - Count_overlap) %>%
  mutate(ratio = sapply(GeneRatio, function(txt) eval(parse(text=txt)), 
                        simplify = TRUE)) %>%
  mutate(`-log_p` = -log10(p.adjust))
go_enrichment_CD4_BP_10 <-  go_enrichment_CD4_BP %>%
  slice_max(order_by = `-log_p`, n = 10) %>%
  mutate(Description = firstup(Description)) %>%
  mutate(Description = stringr::str_wrap(Description, width = 40))

deg_CD8 <- res %>% 
  filter(Cell_Type == "CD8", Method == "DiSC") %>%
  pull(name_sig) %>%
  str_split_1(pattern = "::")
deg_CD8_other_methods <- res %>% 
  filter(Cell_Type == "CD8", Method != "DiSC") %>%
  pull(name_sig) %>% 
  str_c(collapse = "::") %>%
  str_split_1(pattern = "::") %>% 
  unique()
go_enrichment_CD8_BP <- enrichGO(gene = deg_CD8, OrgDb = "org.Hs.eg.db", 
                                 keyType = "SYMBOL", ont = "BP") 
go_enrichment_CD8_BP <- go_enrichment_CD8_BP %>% 
  as_tibble() %>% 
  mutate(Count_overlap = sapply(X = geneID, 
                                FUN = \(x, g_vec){
                                  x = str_split_1(x, pattern = "/")
                                  return(sum(x %in% g_vec))}, 
                                simplify = TRUE, USE.NAMES =FALSE,
                                g_vec = deg_CD8_other_methods),
         Count_DiSC = Count - Count_overlap) %>%
  mutate(ratio = sapply(GeneRatio, function(txt) eval(parse(text=txt)), 
                        simplify = TRUE)) %>%
  mutate(`-log_p` = -log10(p.adjust))
go_enrichment_CD8_BP_10 <-  go_enrichment_CD8_BP %>%
  slice_max(order_by = `-log_p`, n = 10) %>%
  mutate(Description = firstup(Description)) %>%
  mutate(Description = stringr::str_wrap(Description, width = 40))

log_p_comb <- c(go_enrichment_CD4_BP_10$`-log_p`, 
                go_enrichment_CD8_BP_10$`-log_p`)
go_breaks <- seq(from = floor(min(log_p_comb)), to = ceiling(max(log_p_comb)),
                 length.out = 5)
figure4 <- go_enrichment_CD4_BP_10 %>%
  mutate(Description = fct_reorder(Description, `-log_p`, .desc = TRUE)) %>%
  ggplot(aes(x = Description, y = ratio, fill = `-log_p`)) +
  geom_bar(stat = "identity")+
  geom_text(label = go_enrichment_CD4_BP_10$GeneRatio, vjust=-1, size = 2)+
  scale_y_continuous(labels = scales::percent, limits = c(0, 0.11)) + 
  scale_fill_steps(name = "-log10(Adj P)", breaks = go_breaks, 
                   limits = c(min(go_breaks), max(go_breaks)),
                   low = "#bdd7e7", high = "#2171b5")+
  theme_bw()+labs(x = NULL, y = "% of DE Genes Detected")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
figure4
figure5 <- go_enrichment_CD8_BP_10 %>%
  mutate(Description = fct_reorder(Description, `-log_p`, .desc = TRUE)) %>%
  ggplot(aes(x = Description, y = ratio, fill = `-log_p`)) +
  geom_bar(stat = "identity")+
  geom_text(label = go_enrichment_CD8_BP_10$GeneRatio, vjust=-1, size = 2)+
  scale_y_continuous(labels = scales::percent, limits = c(0, 0.11)) + 
  scale_fill_steps(name = "-log10(Adj P)", breaks = go_breaks, 
                   limits = c(min(go_breaks), max(go_breaks)),
                   low = "#bdd7e7", high = "#2171b5")+
  theme_bw()+labs(x = NULL, y = "% of DE Genes Detected")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
figure5
figure4_5 <- lemon::grid_arrange_shared_legend(figure4, figure5, 
                                               position = "right")%>%
  as_ggplot()

figures <- 
  arrangeGrob(figure1,                              
              figure2_3,
              figure4_5, 
              layout_matrix = rbind(c(1,1), c(2,2), c(3,3), c(3, 3))) %>%
  as_ggplot() + 
  draw_plot_label(label = c("(a)", "(b) CD4", "(c) CD8", "(d)", "(e)"), 
                  x = c(0, -0.025, 0.42, 0, 0.45), 
                  y = c(1, 0.75, 0.75, 0.5, 0.5),
                  size = 14, color = "black", 
                  fontface = "plain", family = NULL)
figures
ggsave(filename = "results/covid_pbmc_anal.pdf" , figures,
       width = 8, height = 10)
ggsave(filename = "results/covid_pbmc_anal.jpeg" , figures,
       width = 8, height = 10)
```



# Session info

```{r session info}
sessionInfo()
```

# References

-    Zhang, M., et al. IDEAS: individual level differential expression analysis for single-cell RNA-seq data. Genome Biol 23, 33 (2022). https://doi.org/10.1186/s13059-022-02605-1.
-    Velmeshev, D., et al. Single-cell genomics identifies cell type–specific molecular changes in autism. Science 364. 6441 (2019): 685-689.
-    Stephenson, E., et al. Single-cell multi-omics analysis of the immune response in COVID-19. Nature medicine 27.5 (2021): 904-916.
